clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java

```diff
@@ -59,10 +59,14 @@ public void record(MetricConfig config, double value, long timeMs) {
     @Override
     public double measure(MetricConfig config, long now) {
         double value = stat.measure(config, now);
+        return value / convert(windowSize(config, now));
+    }
+
+    public long windowSize(MetricConfig config, long now) {
         // the elapsed time is always N-1 complete windows plus whatever fraction of the final window is complete
-        long elapsedCurrentWindowMs = now - stat.current(now).lastWindowMs;
+        long elapsedCurrentWindowMs = (now - stat.oldest(now).lastWindowMs) % config.timeWindowMs();
```

Comment:
Take a look at the testcase RateWindowing I added. Is the output time correct? i.e. 75 seconds.

Let me provide an example: Assume we have 3 samples of size 30 seconds each. 
T0 - Start of time
T60 - 60 seconds have elapsed. 2 full sample windows
T75 - A record is called at this time. Note that there was no activity from 60-75.

the current code will produce windowSize of 60 seconds because:
long elaspedCurrentWindowMs = 0 seconds (because the sample just got created). The timestamp used is the create timestamp of the current sample and not the end timestamp of the previous sample. Shouldn't the samples be contiguous?

long elapsedPreviousWindow = (numSamples - 1) * sampleSize = 2 * 30 = 60 seconds

The current code produces this output:
long elapsedPreviousWindow = 60 seconds (same as above)
long elapsedCurrentWindow = 15 seconds. 

Shouldn't this be valid because we actually have 75 seconds of data? IIUC, the current code creates a blackout of 15 seconds.